DIFM Lite: Adds flow for choosing existing sites

aneeshd16 opened the pull request:
### Changes proposed in this Pull Request

P2: pdh1Xd-cu-p2

This change adds a few more screens to the DIFM sign-up flow (/start/do-it-for-me) to allow users to purchase DIFM on existing sites.

### Testing instructions

**Prerequisite**
~Please apply D71573-code on your sandbox.~

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
#### Testing if the new or existing site choice screen is shown for users with existing sites only

1. Log in as a user with 1 or more than 1 existing sites.
2. Start the flow by navigating to `/start/do-it-for-me`. 
3. Confirm that this screen is shown:
![image](https://user-images.githubusercontent.com/5436027/146354468-e092aae0-3719-4703-9873-9e529a7cd738.png)
4. Logout and login again as a user with no existing sites. (Or visit `/start/do-it-for-me` in an incognito window)
5. Confirm that you are directly taken to the Design Picker.

#### Testing the flow as a new user
1. Logout and login again as a user with no existing sites. (Or visit `/start/do-it-for-me` in an incognito window)
2. Go through the flow steps: pick a theme, fill out the site info form and select a domain.
3. Confirm that you are able to complete the purchase.

#### Testing the "New Site" flow as a user with existing sites
1. Login as a user with more than 1 existing site and then navigate to `/start/do-it-for-me`. 
2. On the first step, select "New Site".
3. Go through the flow steps: pick a theme and fill out the site info form.
4. Confirm that the domains step is **not** shown.
5. Confirm that you are able to complete the purchase.

#### Testing the "Existing Site" flow as a user with existing sites
1. Login as a user with more than 1 existing site and then navigate to `/start/do-it-for-me`. 
2. On the first step, select "Existing WordPress.com Site".
3. Confirm that you see the site picker screen:
![image](https://user-images.githubusercontent.com/5436027/146375938-dd97e34e-faf3-4064-a015-f6fc238cda13.png)
4. Confirm that these sites are **not** visible in the site picker: Jetpack/Atomic sites, P2 sites and DIFM in-progress sites.
5. On selecting a site, confirm that the confirmation dialog is shown:
![image](https://user-images.githubusercontent.com/5436027/146376403-470dfaf4-b233-48f5-a31f-522640391f45.png)
6. Confirm that the "Delete site content" button remains disabled till the text input contains the exact site address.
7. Go through the flow steps: pick a theme and fill out the site info form.
8. Confirm that the domains step is **not** shown.
9. Confirm that you are able to complete the purchase.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

aneeshd16 committed:
Adds support for choosing existing sites

aneeshd16 committed:
Send difm data in cart item extra

aneeshd16 committed:
Linting and TS fixes

aneeshd16 committed:
Skip the manage site flow for DIFM

aneeshd16 committed:
Styling fixes

aneeshd16 committed:
Adds explainer text to the delete dialog

aneeshd16 committed:
Copy fix

a8ci18n commented on client/signup/steps/difm-site-picker/index.tsx:
```diff
@@ -0,0 +1,169 @@
+				<p>{ translate( 'The content will be deleted only after the purchase.' ) }</p>
+				<FormLabel htmlFor="confirmDomainChangeInput">
+					{ translate(
+						'Please type in {{warn}}%(siteAddress)s{{/warn}} in the field below to confirm. ' +
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **24** times:
`translate( 'Please type in {{warn}}%(siteAddress)s{{/warn}} in the field below to confirm. Your site will then be gone forever.' )` *ES Score: 9*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/new-or-existing-step-in-difm)

aneeshd16 committed:
Copy and styling changes

aneeshd16 committed:
Minor refactor

a8ci18n commented on client/signup/steps/difm-site-picker/index.tsx:
```diff
@@ -0,0 +1,197 @@
+		return (
+			<Dialog isVisible={ true } buttons={ buttons } onClose={ onCloseDialog }>
+				<h1>{ translate( 'Site Reset Confirmation' ) }</h1>
+				<p>{ translate( 'After completing your purchase:' ) }</p>
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **52** times:
`translate( 'Completing your purchase', { context: 'Loading state on /checkout'} )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/new-or-existing-step-in-difm)

aneeshd16 committed:
Fixes flow for new users

aneeshd16 committed:
Re-enable the check for the delete button

aneeshd16 commented on client/lib/signup/step-actions/index.js:
```diff
@@ -181,10 +179,6 @@ function getNewSiteParams( {
 			site_creation_flow: flowToCheck,
 			timezone_string: guessTimezone(),
 			wpcom_public_coming_soon: 1,
-			difm_lite_site_options: {
```
These have been removed since these options are now passed in the `extra` field of the cart item.

dzver commented on client/signup/steps/new-or-existing-site/index.tsx:
```diff
@@ -0,0 +1,58 @@
+	const subHeaderText = '';
+
+	useEffect( () => {
+		dispatch( saveSignupStep( { stepName: props.stepName } ) );
```
Why do we save the signup step here?

aneeshd16 replied:
Thanks for asking this! I was not entirely sure, and this helped me get more clarity on this. We discussed this briefly in Slack, but I'll add some context here for anyone searching. `saveSignupStep` adds the step to the `signup.progress` slice of the redux store. Saving the step is needed for back functionality to work correctly, especially when skipping a step is allowed.

In this particular case, the only way to move forward from this step is to choose either of the two options (New/Existing site), which calls `submitSignupStep`. `submitSignupStep` also saves the step to the redux store, so calling `saveSignupStep` here is not necessary. However, if we move this step to a different flow or allow skipping this step, we would require `saveSignupStep` here.

dzver reviewed (approved):
Good job!

I think I managed to cover all the cases in the test plan and understood how it works. Left one question

aneeshd16 committed:
Copy updates

aneeshd16 committed:
Adds parameter to opt-out from `isManageSiteFlow`

The DIFM Lite flow deals with both new and existing sites. A user may
choose an existing site for DIFM, press back from checkout and then go
into the new site flow. This satisfies the conditions for setting
`isManageSiteFlow` to `true`. However, when the user then goes through
the new site flow, the `siteId` and `siteSlug` will already be set
causing a breakage - the domains step will be skipped because these
dependencies will already be set. This change adds a new optional
parameter `excludeFromManageSiteFlows` that skips setting `isManageSiteFlow`
to `true`.

aneeshd16 commented:
Rebased.

sirbrillig commented on client/signup/steps/difm-site-picker/index.tsx:
```diff
@@ -0,0 +1,196 @@
+	const { goToNextStep } = props;
+	const [ siteId, setSiteId ] = useState< number | null >( null );
+	const [ confirmDomain, setConfirmDomain ] = useState( '' );
+	const siteDomain = useSelector( ( state ) => getSiteDomain( state, siteId ) );
```
I think this is a TS error: `Argument of type 'number | null' is not assignable to parameter of type 'number'. Type 'null' is not assignable to type 'number'.` Here's a fix:

```suggestion
	const siteDomain = useSelector( ( state ) => getSiteDomain( state, siteId ?? 0 ) );
```

sirbrillig commented on client/signup/steps/difm-site-picker/index.tsx:
```diff
@@ -0,0 +1,196 @@
+
+	const filterSites = ( site: SiteData ) => {
+		return (
+			site.capabilities?.manage_options &&
```
This is also a TS error: `Property 'capabilities' does not exist on type 'SiteData'.`

You might want to add the following to [SiteData](https://github.com/Automattic/wp-calypso/blob/0b4f87169f148a87d4e44a4f868921bc74ab08cc/client/state/ui/selectors/site-data.ts#L3):

```ts
capabilities?: Record< string, boolean >;
```

sirbrillig commented on client/signup/steps/difm-site-picker/index.tsx:
```diff
@@ -0,0 +1,196 @@
+		return (
+			site.capabilities?.manage_options &&
+			! site.jetpack &&
+			! site.options?.is_wpforteams_site &&
```
These two lines are also TS errors which you can fix by adding `is_wpforteams_site: boolean` and `is_difm_lite_in_progress: boolean` to [SiteDataOptions](https://github.com/Automattic/wp-calypso/blob/0b4f87169f148a87d4e44a4f868921bc74ab08cc/client/state/ui/selectors/site-data.ts#L26).

sirbrillig reviewed (commented):
There's a few minor TypeScript errors added by this PR that should be easy to fix.

aneeshd16 committed:
Fixes TS errors

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7057260

Thank you @aneeshd16 for including a screenshot in the description! This is really helpful for our translators.

tyxla commented on client/signup/steps/difm-site-picker/index.tsx:
```diff
@@ -0,0 +1,196 @@
+								warn: <span />,
+							},
+							args: {
+								siteAddress: siteId && siteDomain,
```
All this logic is wrapped in a `if( siteId )` block, so the `sideId &&` check here is unnecessary and can be removed.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7057260).
